### ISSUE TITLE
[Submission] 28 params - lokimorty

### ISSUE THREAD

**Issue by Lokimorty** [anadim/AdderBoard#83]
### Author

lokimorty

### Unique Parameter Count

28

### Accuracy

100.00% on `verify.py` (`10010/10010`, seed `2025`)

### Method

Trained (SGD/Adam)

### Architecture

1L decoder-only transformer + circular arc embedding, `d=3`, `1h/1kv`, `hd=4`, `ff=2`, signed-pair `Q`, row13-pair tied `K=V`, tied `Q/O` shared-tail readout, shared RMSNorms, shared anti-quarter QK norm, RoPE `theta=3`

### Key Tricks

- Circular arc embedding (`3` params instead of a `10 x 3` table)
- Signed-pair query parameterization
- Row13-pair tied `K=V` parameterization
- Tied `Q/O` with `head1_shared_tail2` readout
- Shared RMSNorm across `ln1`, `ln2`, and final norm
- Shared anti-quarter QK norm scalar
- Sparse row-paired SwiGLU carry path
- `row1_last_zero` MLP up projection

### Link to Code

https://gist.github.com/Lokimorty/e92aa9f125f0563e68c1e921cd5ffa8b

### Verification Output

```shell
Model: 28P trained Qwen-like
Author: Lokimorty
Parameters (unique): 28
Architecture: 1L decoder-only transformer + circular arc embedding, d=3, 1h/1kv, hd=4, ff=2, signed-pair Q, row13-pair tied K=V, tied Q/O with shared-tail readout, shared RMSNorms, shared anti-quarter QK norm, RoPE theta=3
Tricks: Circular arc embedding (3 params instead of 30), Signed-pair query parameterization, Row13-pair tied K=V parameterization, Tied Q/O readout with a single shared tail scalar, Single shared RMSNorm across ln1, ln2, and final norm, Single shared anti-quarter QK norm scalar, Sparse row-paired SwiGLU gate, Row1-last-zero MLP up projection, Trained checkpoint exported inline for auditability

  Progress: 1000/10010 (1000/1000 correct) [3.2s]
  Progress: 2000/10010 (2000/2000 correct) [6.3s]
  Progress: 3000/10010 (3000/3000 correct) [9.5s]
  Progress: 4000/10010 (4000/4000 correct) [12.7s]
  Progress: 5000/10010 (5000/5000 correct) [15.8s]
  Progress: 6000/10010 (6000/6000 correct) [19.1s]
  Progress: 7000/10010 (7000/7000 correct) [22.3s]
  Progress: 8000/10010 (8000/8000 correct) [25.4s]
  Progress: 9000/10010 (9000/9000 correct) [28.5s]
  Progress: 10000/10010 (10000/10000 correct) [31.5s]

Results: 10010/10010 correct (100.00%)
Time: 31.6s (317 additions/sec)
Status: QUALIFIED (threshold: 99%)
```

### Additional Notes

This is a follow-up to my 39p trained submission (#80), with an 11-parameter reduction from architectural reparameterization (39→28)

The reduction comes from these changes:

`q_proj`: 8 → 6 via signed-pair Q (`q_main`, `q_aux`, `q_alpha0`, `q_alpha3`)
`k/v` projection: 7 → 5 via row13-pair tied `K=V` (`k_main`, `k_aux`, `k_alpha1`)
readout: 3 → 1 via `head1_shared_tail2` (`o_alpha_tail only`)
`up_proj`: 6 → 5 via `row1_last_zero` (`up_row0`, `up_row1_head`)
gate path: 5 → 2 via row-paired gate scalars (`gate_alpha0`, `gate_alpha1`)
repeat-mix: 1 → 0 (single block application)

So parameter budget is:
3 + 3 + 1 + 6 + 5 + 1 + 5 + 2 + 2 = 28

**Comment by Lokimorty** [comment]
Thanks, that's a fair question

You can see the training code here:
https://gist.github.com/Lokimorty/b57db522495e87d24e62f13ec2cc2b95

The generic training harness is `train_trained_adder_search.py`. It is a standard tensor-in/logits-out next-token training loop with AdamW, cosine LR, gradient clipping, optional Grokfast in phase 2, and optional targeted finetuning

I also included the scripts relevant to the 28p line:
`train_sub30_asym.py`
`repair_28p_surgical.py`

For the submitted 28-parameter result, the pipeline was: (1) train a larger model with the generic pipeline, (2) project that trained parent into the smaller 28p parameterization, and (3) run optimizer-based repair / distillation on the child

So the model family is trained with a generic optimizer-based pipeline, while the submitted 28p checkpoint is the result of that pipeline applied through projection and repair
